openapi.yml

```diff
@@ -12,6 +12,14 @@ paths:
       responses:
         201:
           description: Order has been accepted.
+        400:
+          description: Gas-tip is not sufficient
+        401:
+          description: Unauthorized, your account is deny-listed
+        409:
+          description: Order exists already
+        429:
+          description: Too many order placements
```

Comment:
These status codes have fairly specific meaning for the http standard but I don't think they make sense to distinguish application level errors. Maybe 409 and 429 do make sense but I don't think this has to be part of the api guarantees.

We should specify in this in the api only for errors the frontend wants to be able to distinguish and react to. Instead of status codes it is probably better to introduce something like an error enum or json which we return as the body for 400 errors.

Reply:
>These status codes have fairly specific meaning for the http standard but I don't think they make sense to distinguish application level errors. Maybe 409 and 429 do make sense but I don't think this has to be part of the api guarantees.

you are right, I read more about it and agree!

> json which we return as the body for 400 errors.

Yeah, I like it.

>the frontend wants to be able to distinguish and react to.
I also agree! I thought that the might want to react to all of them. Let's wait for Anxo's response.

Reply:
I would be careful not to conflate HTTP errors with API errors. That being said, I think the error codes are in general well chosen. Some proposed changes:
- 
```suggestion
        400:
          description: Order is malformed
        403:
          description: Forbidden, your account is deny-listed
        429:
          description: Too many order placements
        500:
          description: Error adding an order
```

So, the reason I would suggest a `403` instead of a `401` is that the former is intended to indicate that a certain authenticated user (authenticated via signature in this case) does not have sufficient access (i.e. is not authorized), while the latter indicates that authentication is required but missing (which is not the case since a signature is a valid authentication). From the [MDN page](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/403):
> This status is similar to 401, but in this case, re-authenticating will make no difference. The access is permanently forbidden and tied to the application logic, such as insufficient rights to a resource.

> Maybe 409 and 429 do make sense but I don't think this has to be part of the api guarantees.

I like 429 as this is generally used for rate limiting and is a pretty standard code for this scenario.

> return as the body for 400 errors.

I would reserve 400 for malformed input (such as the order JSON is missing properties or the signature is invalid) and use 500 errors for general errors adding orders (duplicate entries, missing balance, etc.)

In fact, to make it easy for the FE to differentiate errors, we can always return a
```rust
struct Error {
  code: ErrorKind,
  message: String
}
#[repr(i32)]
enum ErrorKind {
  MissingOrderData,
  InvalidSignature,
  DuplicateOrder,
  InsufficientFunds,
}
```

Which would serialize to something like:
```json
{
  "code": 2, // ErrorKind::DuplicateOrder
  "message": "Order exists already"  
}
```

Reply:
Not wanting to argue over which http error fits better is one of the reasons I don't want to use http errors and prefer the return body enum. We know which errors we have so let's represent those in the most obvious way possible like in your json struct and not think about the http status code as much. For me it is enough to say: 400 = couldn't understand (deserialize) request, 500 = any other error. And then for 500 we return this json response to describe the error.

I would make the error enum be represented as a struct with an openapi enum (which is json string) and and optional message that has more detail. No need to convert the enum to a number first.

Reply:
> And then for 500 we return this json response to describe the error.

I would return it for both `500` and `400` errors. There are things like invalid signatures that are orthogonal to deserialization but would, IMO constitute a "Bad Request" which would be nice to get a more descriptive error on.

> I would make the error enum be represented as a struct with an openapi enum (which is json string)

Yeah, that is much better.

Reply:
I agree with most of everything said in this thead

I like the proposal of error codes of Nick, I think is not only useful for the front end, you'll find it useful for Prometheus stats for example. 

Anyways, the way I would define this errors is, by setting an generic structure for errors. So everything that is not a `200` should have at least a "description" (string) and a "type" (enum). 

Normally, most errors would suffice with that data. But in some cases, you might want to add additional data.
Not required for us, but maybe convenient for you, to include the `stacktrace` in the error `500`. Although not sure strictly pretty, just for saving you the time to go and check the logs and match it with a certain time.

Reply:
> Normally, most errors would suffice with that data

Structured data for certain error types is a really nice idea! 

> Not required for us, but maybe convenient for you, to include the stacktrace in the error 500

My only concern with that might accidentally leaking private information (for example something part of the configuration).

Reply:
> Not wanting to argue over which http error fits better is one of the reasons

One thing worth mentioning is that certain HTTP errors do affect browser behaviour. For example a 401 can trigger a login prompt.

Reply:
> My only concern is that it might accidentally leaking private information (for example something from the configuration).

Fair enough. I've seen many times settling a flag just for this, in dev environments. There could be also a leak but not so important.

> One thing worth mentioning is that certain HTTP errors do affect browser behaviour. For example a 401 can trigger a login prompt.

mm, I've never saw it just by sending a 401, I think normally it's triggered by a special header the serve sends, signaling they want a basic authentication. Since the browsers knows how to do it, they just go for it. 
Anyway, it was a good call you changing the 401 for a 403. One is authentication, the other authorization.

Reply:
> mm, I've never saw it just by sending a 401, I think normally it's triggered by a special header the serve

:see_no_evil: Yes, it needs a `WWW-Authenticate` header.